kratos/geometries/geometry_shape_function_container.h

```diff
@@ -60,25 +60,38 @@ class GeometryShapeFunctionContainer
     /// Pointer definition of GeometryShapeFunctionContainer
     KRATOS_CLASS_POINTER_DEFINITION( GeometryShapeFunctionContainer );
 
+    /// Defining enum of integration method
     using IntegrationMethod = TIntegrationMethodType;
 
+    /// Size types
     typedef std::size_t IndexType;
     typedef std::size_t SizeType;
 
+    /// Integration points
     typedef IntegrationPoint<3> IntegrationPointType;
     typedef std::vector<IntegrationPointType> IntegrationPointsArrayType;
     typedef std::array<IntegrationPointsArrayType, IntegrationMethod::NumberOfIntegrationMethods> IntegrationPointsContainerType;
 
-
+    /// Shape functions
     typedef std::array<Matrix, IntegrationMethod::NumberOfIntegrationMethods> ShapeFunctionsValuesContainerType;
-    typedef std::array<DenseVector<Matrix>, IntegrationMethod::NumberOfIntegrationMethods> ShapeFunctionsLocalGradientsContainerType;
 
+    /// First derivatives/ gradients
     typedef DenseVector<Matrix> ShapeFunctionsGradientsType;
+    typedef std::array<DenseVector<Matrix>, IntegrationMethod::NumberOfIntegrationMethods> ShapeFunctionsLocalGradientsContainerType;
+
+    /// Higher order derivatives
+    typedef DenseVector<Matrix>
+        ShapeFunctionsDerivativesType;
+    typedef DenseVector<DenseVector<ShapeFunctionsDerivativesType> >
+        ShapeFunctionsDerivativesIntegrationPointArrayType;
+    typedef std::array<ShapeFunctionsDerivativesIntegrationPointArrayType, IntegrationMethod::NumberOfIntegrationMethods>
+        ShapeFunctionsDerivativesContainerType;
 
     ///@}
     ///@name Life Cycle
     ///@{
 
+    /// Constructor for single integration point having the full containers
     GeometryShapeFunctionContainer(
         IntegrationMethod ThisDefaultMethod,
```

Comment:
I see that you have one container for all of the integration points at once.

My idea was to have one different container per each integration point.

so that you could do something like

~~~
auto integraiton_point_data = all_integration_points_container.GetPoint( 5 )
array_1d<double,3> v = data.GetVector(derivative_order,  .... ) <-------> this is what you would use in the PR about o
double d = data.GetValue(derivative_order, component ......)
~~~

Reply:
This is only piecewise possible. However, this is not backwards compatible. The herein posted solution is applicable for all kratos.

Reply:
ok